### ISSUE TITLE
Can't remove cmd from application without ugly hack

### ISSUE THREAD

**Issue by philipnrmn** [mesosphere/marathon#2147]
- Create a simple application eg `cmd: "sleep 10"`
- Update the config with a docker image and remove the `cmd` field
- The `cmd` field persists if not set or set to null (as expected), and responds with a `minLength` validation failure if set to blank. 

The UI currently works around this by sending a single space character.

**Comment by kolloch** [comment]
This could be solved by the usual `PUT` semantic (#1418),  related documentation issue #1822.

An alternative solution would be to allow the syntax `"fieldName": null` to explicitly remove that field from an app definition.

**Comment by roobert** [comment]
This is a pretty brutal bug and tripped me up, as a first time marathon user I wouldn't have expected the marathon-ui to transparently override my explicitly configured docker images. Ouch.

**Comment by philipnrmn** [comment]
Hi @roobert, the UI won't override an explicitly configured docker image. There was a bug along those lines - #2338 - which was resolved in Marathon 0.11.1. Which version of Marathon are you using?

**Comment by roobert** [comment]
Hi @phildougherty, here are the versions I'm using:

```
# dpkg -l | grep -E 'mesos|marathon' 
ii  marathon                        0.11.1-1.0.432.debian81           amd64        Cluster-wide init and control system for services running on Apache Mesos
ii  mesos                           0.25.0-0.2.70.debian81            amd64        Cluster resource manager with efficient resource isolation
```

My Dockerfile: https://github.com/roobert/webapp-hostname/blob/master/Dockerfile

Manual test, no args:
```
# docker run --rm -ti webapp-hostname
WARNING: Your kernel does not support memory swappiness capabilities, memory swappiness discarded.
[2015-11-04 17:10:45] INFO  WEBrick 1.3.1
[2015-11-04 17:10:45] INFO  ruby 2.1.5 (2014-11-13) [x86_64-linux-gnu]
== Sinatra (v1.4.6) has taken the stage on 4567 for development with backup from WEBrick
[2015-11-04 17:10:45] INFO  WEBrick::HTTPServer#start: pid=1 port=4567
^C== Sinatra has ended his set (crowd applauds)
[2015-11-04 17:10:49] INFO  going to shutdown ...
[2015-11-04 17:10:49] INFO  WEBrick::HTTPServer#start done.
#
```

Extra args to set alternate port:
```
# docker run --rm -ti webapp-hostname 5555
WARNING: Your kernel does not support memory swappiness capabilities, memory swappiness discarded.
[2015-11-04 17:11:10] INFO  WEBrick 1.3.1
[2015-11-04 17:11:10] INFO  ruby 2.1.5 (2014-11-13) [x86_64-linux-gnu]
== Sinatra (v1.4.6) has taken the stage on 5555 for development with backup from WEBrick
[2015-11-04 17:11:10] INFO  WEBrick::HTTPServer#start: pid=1 port=5555
^C== Sinatra has ended his set (crowd applauds)
[2015-11-04 17:11:12] INFO  going to shutdown ...
[2015-11-04 17:11:12] INFO  WEBrick::HTTPServer#start done.
#
```

If I try and launch this image through the marathon UI by setting:

* ID = test0
* Memory = 1024
* Image = webapp-hostname

And I leave the Command unset, then launch the instance, status shows WAITING then DELAYED.

Config:
```
{
  "type": "DOCKER",
  "volumes": [],
  "docker": {
    "image": "webapp-hostname",
    "network": "HOST",
    "privileged": false,
    "parameters": [],
    "forcePullImage": false
  }
}
```

Note: I tried both BRIDGE and HOST network types.

I repeated the test with:

* Command = 5555

..and the same situation occurred.

If I set:

* Command = while :; do echo -n; done

Then the container launches immediately:

```
# docker ps -a
CONTAINER ID        IMAGE                       COMMAND                   CREATED             STATUS              PORTS                     NAMES
c69d1c207081        webapp-hostname             "/bin/sh -c 'while :;"    29 seconds ago      Up 29 seconds                                 mesos-264c44fe-8cc4-46d7-94e1-76dbae2930fa-S0.ad847202-0954-4929-997e-bb3077eeb885

# docker exec -ti c69d1c207081 /bin/bash
root@c69d1c207081:/# ps -ax
  PID TTY      STAT   TIME COMMAND
    1 ?        Rs     0:50 /bin/sh -c while :; do echo -n; done
    5 ?        Ss     0:00 /bin/bash
   10 ?        R+     0:00 ps -ax
```

Config looks the same:
```
{
  "type": "DOCKER",
  "volumes": [],
  "docker": {
    "image": "webapp-hostname",
    "network": "BRIDGE",
    "privileged": false,
    "parameters": [],
    "forcePullImage": false
  }
}
```

The only difference is that Command was set, which overrode the ENTRYPOINT in my Dockerfile.

According to the docker docs, this isn't the expected behavior: http://docs.docker.com/engine/reference/run/#cmd-default-command-or-options

If I use the some JSON and the API then the instance launches *almost* as expected: https://github.com/roobert/webapp-hostname/blob/master/marathon-webapp-hostname.json

Config from the UI:

```
{
  "type": "DOCKER",
  "volumes": [],
  "docker": {
    "image": "webapp-hostname",
    "network": "BRIDGE",
    "portMappings": [
      {
        "containerPort": 4444,
        "hostPort": 0,
        "servicePort": 10000,
        "protocol": "tcp"
      }
    ],
    "privileged": false,
    "parameters": [],
    "forcePullImage": false
  }
}
```

```
# docker logs 6d12afa2a903              
[2015-11-03 16:21:56] INFO  WEBrick 1.3.1
[2015-11-03 16:21:56] INFO  ruby 2.1.5 (2014-11-13) [x86_64-linux-gnu]
== Sinatra (v1.4.6) has taken the stage on 31369 for development with backup from WEBrick
[2015-11-03 16:21:56] INFO  WEBrick::HTTPServer#start: pid=1 port=31369

# docker exec 6d12afa2a903 ps -ax  
  PID TTY      STAT   TIME COMMAND
    1 ?        Ssl    0:05 /app.rb
   75 ?        Rs     0:00 ps -ax
```

Command is 'Unspecified'

If I run `docker ps -a` straight after launching an instance from the UI I can see:

```
c76fc1058e68        webapp-hostname             "/bin/sh -c ' '"         5 seconds ago       Exited (0) 3 seconds ago                       mesos-264c44fe-8cc4-46d7-94e1-76dbae2930fa-S0.b613f321-c54d-4207-adbd-17430c7107c8
```

So it looks like this feature/bug still exists in 0.11.1.

Any help appreciated, thanks.

**Comment by magcius** [comment]
I'm on version 0.11.1 and just hit this bug as well. It seems creating applications through the web UI is just completely broken.

As a workaround, perhaps an empty string or entirely whitespace in an application description should be seen as an empty value?

**Comment by magcius** [comment]
In fact, the bug seems to be extremely obvious: https://github.com/mesosphere/marathon-ui/blob/v0.11.1/src/js/components/modals/AppModalComponent.jsx#L204-L208

When creating an application, if the "docker image" field is non-empty and the cmd field is empty, it's overwritten with a single space, meaning that the entrypoint is never used.

**Comment by roobert** [comment]
@philipnrmn, any update on this? Cheers!

**Comment by philipnrmn** [comment]
Hi @magcius, the reason for those lines of code existing is that when *editing* an application definition and *adding* a docker image field and *removing* a cmd field, the cmd needs to be overwritten with a single space. Sending `null` or `""` used to leave the old cmd intact. Recent versions of the Marathon API have fixed this issue.

**Comment by roobert** [comment]
Hi @philipnrmn, thanks for the reply but what version of the marathon-api is this fixed in? As described in my earlier post I'm using 0.11.1 and this bug still exists.

Thanks!

**Comment by philipnrmn** [comment]
Hi @roobert, sorry I missed your detailed feedback upthread! I think you tagged the wrong phil. 

Re your immediate comment, the API issue is fixed in 0.13, which is set for imminent release - see [this commit](https://github.com/mesosphere/marathon/commit/161d4aebdf917c653fdc46ee15d6dd9241ca2b0c). I'll do some manual testing now on 0.13 and master, and post an update shortly.

**Comment by roobert** [comment]
@philipnrmn, Great, thanks again! My fault for tagging the wrong Phil :)

**Comment by magcius** [comment]
@philipnrmn I can't see any evidence that code is ran only for editing and not for adding -- it seems it's ran as a post-processing step after every click of the Submit button. I haven't tested 0.13 yet, but on 0.11.1, I can confirm that the bug exists when simply adding an application.

I'm sure 0.13 has some excellent improvements, but it remains that you cannot create any usable Docker application using the Marathon UI because of this bug. That seems like a pretty major issue to have in a stable release.

And do you have a reference for the API issue being fixed? #1418 remains open and in the backlog milestone, and the linked commit simply goes to a few test suite changes.

Thanks.

**Comment by magcius** [comment]
Actually, I'm a bit curious now. What's the value in replacing `cmd` with a single space? On 0.11.1, this causes Marathon to think it should run `/bin/sh -c " "`, which will complete instantly, causing the task to spin up and forever down.

Is there a version of Marathon where a cmd of a single space will have usable behavior? In 0.11.1, any non-null value is treated as a shell command: https://github.com/mesosphere/marathon/blob/v0.11.1/src/main/scala/mesosphere/mesos/TaskBuilder.scala#L225-L230

**Comment by philipnrmn** [comment]
Creating a docker image in 0.13.0-RC6 without specifying a command does not cause a `cmd` field to be set with any value. If this is an issue in 0.11.1, we can say it's fixed in 0.13.0.

**Comment by philipnrmn** [comment]
Also in 0.13.0-RC6:
Creating a non-dockerised application, then editing it by adding a docker image and removing the command results in the field `"cmd": " "` being sent to the API.

**Comment by philipnrmn** [comment]
In 0.11.1, creating a dockerised application without specifying a command does result in the field `"cmd": " "` being sent to the API.

**Comment by philipnrmn** [comment]
@magcius, thanks for the extra feedback. You're quite right, the API has in fact not been fixed (I updated my earlier comment to reflect this) and sending a space is a bad solution which causes the subsequent application to be unusable. aegrescit medendo. 

Just wanted to clarify this point:
> you cannot create any usable Docker application using the Marathon UI because of this bug

In 0.11.1, you cannot create a working Docker application without specifying a command because of this bug. The workaround is to copy the command or entrypoint from your Dockerfile into the `cmd` field of the application. This is fixed in 0.13.0.

**Comment by philipnrmn** [comment]
On the matter of the API fix, I've just done some testing, and [the commit referenced above](https://github.com/janisz/marathon/commit/3ec38e82f590896fe007ece9ee07486bde8dfaf5) does indeed fix this bug. It is fixed in the API as of 0.13.0, which was coincidentally announced a few moments ago. 

@roobert huge thanks for your very detailed logs, it made testing this about a million times easier than it would otherwise have been. 

@magcius #1418 is a broader story around PUT semantics, which isn't yet resolved. However, I can confirm that this specific bug is now fixed.

**Comment by roobert** [comment]
@philipnrmn glad I could help, thanks for the awesome project!
